src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/DeclarationTests.cs

```diff
@@ -1577,5 +1586,398 @@ private static void CompileDeclaration(EvaluationContext context, string declara
+                testData.GetMethodData("<>x.<>m0<T>").VerifyIL(
+    @"{
+  // Code size       58 (0x3a)
+  .maxstack  4
+  .locals init (object V_0, //y
+                bool V_1,
+                object V_2,
+                System.Guid V_3,
+                object V_4)
+  IL_0000:  ldtoken    ""object""
+  IL_0005:  call       ""System.Type System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)""
+  IL_000a:  ldstr      ""i""
+  IL_000f:  ldloca.s   V_3
+  IL_0011:  initobj    ""System.Guid""
+  IL_0017:  ldloc.3
+  IL_0018:  ldnull
+  IL_0019:  call       ""void Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.CreateVariable(System.Type, string, System.Guid, byte[])""
+  IL_001e:  ldstr      ""i""
+  IL_0023:  call       ""object Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetVariableAddress<object>(string)""
+  IL_0028:  ldarg.0
+  IL_0029:  dup
+  IL_002a:  stloc.s    V_4
+  IL_002c:  stind.ref
+  IL_002d:  ldloc.s    V_4
+  IL_002f:  ldnull
+  IL_0030:  cgt.un
+  IL_0032:  call       ""object C.Test(bool)""
+  IL_0037:  starg.s    V_0
+  IL_0039:  ret
+}");
+            });
+        }
+
+        [Fact]
+        public void PatternLocals_Assignment_04()
+        {
+            var source =
+@"class C
+{
+    static object F;
+    static void M<T>(int x)
+    {
+        object y;
+        if (x == 1)
+        {
+            object z;
+        }
+    }
+
+    static int Test(bool x)
+    {
+        return 1;
+    }
+}";
+            var compilation0 = CreateCompilationWithMscorlib(source, options: TestOptions.DebugDll);
+
+            WithRuntimeInstance(compilation0, runtime =>
+            {
+                var context = CreateMethodContext(runtime, "C.M");
+
+                CompilationTestData testData;
+                string error;
+                testData = new CompilationTestData();
+                context.CompileAssignment("x", "Test(x is int i)", out error, testData);
+                testData.GetMethodData("<>x.<>m0<T>").VerifyIL(
+    @"{
+  // Code size       51 (0x33)
+  .maxstack  4
+  .locals init (object V_0, //y
+                bool V_1,
+                object V_2,
+                System.Guid V_3)
+  IL_0000:  ldtoken    ""int""
+  IL_0005:  call       ""System.Type System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)""
+  IL_000a:  ldstr      ""i""
+  IL_000f:  ldloca.s   V_3
+  IL_0011:  initobj    ""System.Guid""
+  IL_0017:  ldloc.3
+  IL_0018:  ldnull
+  IL_0019:  call       ""void Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.CreateVariable(System.Type, string, System.Guid, byte[])""
+  IL_001e:  ldstr      ""i""
+  IL_0023:  call       ""int Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetVariableAddress<int>(string)""
+  IL_0028:  ldarg.0
+  IL_0029:  stind.i4
+  IL_002a:  ldc.i4.1
+  IL_002b:  call       ""int C.Test(bool)""
+  IL_0030:  starg.s    V_0
+  IL_0032:  ret
+}");
+            });
+        }
+
+        [Fact]
+        public void PatternLocals_Assignment_05()
```

Comment:
Each one tests a different path through lowering.

---
In reply to: [77868272](https://github.com/dotnet/roslyn/pull/13535#discussion_r77868272) [](ancestors = 77868272)